### DISCUSSION TITLE
Seeking Help: Implementing a Horizontal ListBox with Variable Item Widths in AvaloniaUI

### DISCUSSION THREAD

**Discussion by xgirones** [AvaloniaUI/Avalonia#11609] [Q&A] [answered]
Hello AvaloniaUI community! I'm new to Avalonia and I'm currently working on implementing a ListBox that arranges its items horizontally, with each item having a potentially different width. Unfortunately, despite my attempts, all the items in the ListBox are currently displaying with the same width, resulting in an aesthetically unpleasing layout. I've spent several hours searching online for a solution, but haven't been successful so far. Could anyone kindly guide me on what I might be doing wrong and offer suggestions on how to resolve this issue? Your assistance would be greatly appreciated!

This is the code I tried:

 ```
<ListBox Name="MyLB" Height="160" Margin="2,5,2,0" VerticalAlignment="Top"
             ScrollViewer.HorizontalScrollBarVisibility="Visible">
      <ListBox.Styles>
        <Style Selector="ListBoxItem">
          <Setter Property="Focusable" Value="True"/>
          <Setter Property="IsHitTestVisible" Value="False"/>
        </Style>
      </ListBox.Styles>    
      <ListBox.ItemsPanel>
        <ItemsPanelTemplate>
          <StackPanel Orientation="Horizontal"/>
        </ItemsPanelTemplate>
      </ListBox.ItemsPanel>
      <ListBox.ItemTemplate>
        <DataTemplate DataType="uc:MyItemControl"/>
      </ListBox.ItemTemplate>
    </ListBox>
```

_MyItemControl_ is defined as follows
```
	<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:vm="using:MyProject.ViewModels"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="MyProject.Views.MyItemControl"
             x:DataType="vmMyItemViewModel">
  <Grid>
    <Grid.RowDefinitions>
      <RowDefinition Height="Auto"/>
      <RowDefinition Height="Auto"/>
    </Grid.RowDefinitions>
    <Image Name="ItemImage" Grid.Row="0" Margin="2" VerticalAlignment="Stretch"
           Stretch="UniformToFill"/>
    <Label Name="ItemLabel" Grid.Row="1" FontSize="14" FontWeight="SemiBold"
           Margin="2" Width="NaN" Height="NaN" HorizontalAlignment="Center" VerticalAlignment="Center"/>
  </Grid>
</UserControl>
```
_ItemImage_ and _ItemLabel_ are bound using ReactiveUI:

```
    this.WhenActivated(disposables =>
    {
        this.OneWayBind(this.ViewModel, vm => vm.Image, view => view.ItemImage.Source)
            .DisposeWith(disposables);

        this.OneWayBind(this.ViewModel, vm => vm.Text, view => view.ItemLabel.Content)
            .DisposeWith(disposables);
    });
```

**Comment by timunie** [comment]
Which Avalonia version?

**Comment by xgirones** [comment]
Avalonia 0.10.21

**Comment by timunie** [comment, marked as answer]
I think your DataTemplate is wrong. Seems to work for me: 
![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/df0fed54-0735-4377-88e3-86c3fefa7e14)

```xml
<ListBox.ItemTemplate>
    <DataTemplate DataType="model:MyItemModel">
        <views:MyItemControl/>
    </DataTemplate>
</ListBox.ItemTemplate>
```

**Reply by xgirones** [reply]
Thank you for your suggestion! I made your recommended changes and now everything works perfectly. However, I encountered another issue with the bitmap display. There was a blank margin on the right side, which I discovered was caused by the resolution parameter. I initially set the resolution to 72 dpi when creating the bitmap from an Emgu.CV Mat object. However, any resolution other than 96 dpi leads to improper rendering of the bitmap. I expected the ListBox presentation to be independent of the bitmap resolution. Is this possibly a bug?

```
                    bms = new Bitmap(PixelFormat.Bgra8888,
                                     AlphaFormat.Opaque,
                                     imgWarped.Mat.DataPointer,
                                     new PixelSize(imgWarped.Width, imgWarped.Height),
                                     new Avalonia.Vector(96, 96), // Anything different from 96 dpi renders improperly.
                                     imgWarped.Mat.Step);
```

**Reply by timunie** [reply]
try it in a blank 11.0-rc1 app. Maybe changed there.

**Reply by xgirones** [reply]
Thank you! I'll give version 11.0-rc1 a try as soon as I have some free time. Up to now, I've successfully managed to get my desktop application running on both Windows and Linux platforms. I must say, AvaloniaUI has proven to be quite impressive in terms of performance and functionality.
